Python/executor.c

```diff
@@ -140,7 +140,7 @@ _PyUopExecute(_PyExecutorObject *executor, _PyInterpreterFrame *frame, PyObject
     // This presumes nothing was popped from the stack (nor pushed).
     DPRINTF(2, "DEOPT: [Opcode %d, operand %" PRIu64 "]\n", opcode, operand);
     OPT_HIST(trace_uop_execution_counter, trace_run_length_hist);
-    frame->prev_instr--;  // Back up to just before destination
+    frame->instr_ptr--; // Back up to just before destination
```

Comment:
To make `-Xuops` work, this line should be deleted. With it things go wrong quickly (it backs up to the instruction just *before* the instruction being deoptimized). Without it things eventually still go wrong, I have to look more into that.

Reply:
(Mostly notes to self:) I'm still debugging this.

It looks like there's a trace that ends in a `_PUSH_FRAME` expanded from `CALL_PY_EXACT_ARGS 0` to a generator. The frame ends there with `_EXIT_TRACE` because the function version cache doesn't have an entry -- this feels like the smoking gun. So the generator executes `COPY_FREE_VARS` and then `RETURN_GENERATOR`, and returns to the Tier 1 (bytecode) interpreter. I've checked the debug output from iterations before the Tier 2 interpreter was involved, and there it returns to a `CALL_METHOD_DESCRIPTOR_O 1` opcode (specialized from `CALL 1`). But on the iteration initiated by Tier 2 it seems to return to the unoptimized `CALL_PY_EXACT_ARGS 0` instruction.

So it looks like after the first call we're jumping back to the same call -- this would imply that `frame->next_instr_offset == 0` at a point where we didn't expect it to be. Could it be that the removal of the line `frame->next_instr_offset = next_instr - frame->instr_ptr;` from `_PUSH_FRAME` by your last commit was wrong? Not really, since that removal caused `_PUSH_FRAME` to be disqualified from being a valid uop. But maybe something else is supposed to set that variable.

I think I have to review how generate_cases.py special-cases `_SAVE_CURRENT_IP`. Later.

Reply:
I made some progress here, but not enough. I concocted an elaborate scheme to undo part of Mark's gh-110755, to make `_SAVE_CURRENT_IP` set `next_instr_offset` to `oparg`. This almost works, but when the Tier 2 interpreter returns, the `ENTER_EXECUTOR` code jumps to label `resume_frame` (in ceval.c), which... **resets** `next_instr_offset`. :-( Since there are other places that jump to the same label, I can't just not reset it. I think. I will continue my investigation another day. (But you definitely need to lose the `frame->instr_ptr--;` from executor.c that I previously flagged in a comment.

Reply:
I have a candidate in https://github.com/python/cpython/pull/111144. Waiting for tests.

Reply:
I am happy now; you should be able to squash-merge https://github.com/iritkatriel/cpython/pull/55 here.